### PR TITLE
Made changes to support elasticsearch 5

### PR DESCRIPTION
**Pull request by Matityahul**

### CODE REVIEW

File: tornadoes/__init__.py [hunk 1]
```diff
@@ -10,62 +10,66 @@
 
 
 class ESConnection(object):
+    _MATCH_ALL_QUERY = {"query": {"match_all": {}}}
```
Comment (sxtomerg) [inline tornadoes/__init__.py:13]: First time someone name this query. nice.. :)

File: tornadoes/__init__.py [hunk 2]
```diff
@@ -10,62 +10,66 @@
...
 
-    def __init__(self, host='localhost', port='9200', io_loop=None, protocol='http', custom_client=None):
+    # TODO : timeout, max_retries, retry_on_timeout
```
Comment (sxtomerg) [inline tornadoes/__init__.py]: TODO?

File: tornadoes/__init__.py [hunk 3]
```diff
@@ -10,62 +10,66 @@
...
-        index = kwargs.pop('index', '_all')
-        doc_type = '/%s' % kwargs.pop('type', '')
+    @staticmethod
+    def create_path(method, index='_all', type='', **kwargs):
+        query_string = ESConnection._create_query_string(kwargs)
 
-        parameters = {}
-        for param, value in kwargs.items():
-            parameters[param] = value
+        path = "/%(index)s/%(type)s/_%(method)s" % {"method": method, "index": index, "type": type}
```
Comment (sxtomerg) [inline tornadoes/__init__.py:52]: You have the words (method, index, type) defined 3 times in the same line (TERSE 👎 ) :)

what about "/{}/{}/{}".format(method, index, type)

File: tornadoes/__init__.py [hunk 4]
```diff
@@ -10,62 +10,66 @@
...
         self.bulk = BulkList()
         self.client = custom_client or AsyncHTTPClient(self.io_loop)
-        self.httprequest_kwargs = {}     # extra kwargs passed to tornado's HTTPRequest class e.g. request_timeout
+
+        # extra kwargs passed to tornado's HTTPRequest class e.g. request_timeout
+        self.http_request_kwargs = http_request_kwargs or {}
 
     @staticmethod
-    def from_uri(uri, io_loop=None, custom_client=None):
+    def _create_query_string(params):
```
Comment (sxtomerg) [inline tornadoes/__init__.py:26]: This method should be beneath "from_uri" and "request_document"

File: tornadoes/__init__.py [hunk 5]
```diff
@@ -10,62 +10,66 @@
...
-            'custom_client': custom_client
-        })
+        return cls(host=parsed.hostname, protocol=parsed.scheme, port=parsed.port, io_loop=io_loop,
+                   custom_client=custom_client, http_request_kwargs=http_request_kwargs)
 
-    def create_path(self, method, **kwargs):
-        index = kwargs.pop('index', '_all')
-        doc_type = '/%s' % kwargs.pop('type', '')
+    @staticmethod
+    def create_path(method, index='_all', type='', **kwargs):
```
Comment (sxtomerg) [inline tornadoes/__init__.py:49]: "method_name" better describes the parameter

File: tornadoes/__init__.py [hunk 6]
```diff
@@ -112,30 +116,31 @@ def delete(self, index, type, uid, parameters=None, callback=None):
...
         if source:
             source = json_encode(source)
 
         self.post_by_path(path=path, callback=callback, source=source)
 
     def request_document(self, index, type, uid, method="GET", body=None, parameters=None, callback=None):
+        query_string = ESConnection._create_query_string(parameters)
+
         path = '/{index}/{type}/{uid}'.format(**locals())
         url = '%(url)s%(path)s?%(querystring)s' % {
```
Comment (sxtomerg) [inline tornadoes/__init__.py:137]: url, path, querystring is written 3 times :)
I know it's not your code :P

File: tornadoes/tests/unit/test_tornadoes.py [hunk 7]
```diff
@@ -170,26 +172,24 @@ def test_count_specific_query(self):
...
+    def test_count_specific_query_with_illegal_parameters(self):
         source = {"query": {"term": {"_id": "171171"}}}
         source = self._set_count_query(source)
         parameters = {'refresh': True}
         self.es_connection.count(callback=self.stop, source=source, parameters=parameters)
         response = self.wait()
-        response_dict = self._verify_response_and_returns_dict(response)
-        self.assertEqual(response_dict["count"], 1)
-        self.assertTrue(response.request.url.endswith('_count?refresh=True'))
+        if self.version < 5:
```
Comment (sxtomerg) [inline tornadoes/tests/unit/test_tornadoes.py:181]: I don't know where you are getting the version parameter, 
but it seems that you should know in the test which version it is and assert according to it.
Maybe separate to 2 unit tests with two different versions

File: tornadoes/__init__.py [hunk 8]
```diff
@@ -10,62 +10,66 @@
...
+    def __init__(self, host='localhost', port='9200', io_loop=None, protocol='http', custom_client=None,
+                 http_request_kwargs=None):
         self.io_loop = io_loop or IOLoop.instance()
         self.url = "%(protocol)s://%(host)s:%(port)s" % {"protocol": protocol, "host": host, "port": port}
         self.bulk = BulkList()
         self.client = custom_client or AsyncHTTPClient(self.io_loop)
-        self.httprequest_kwargs = {}     # extra kwargs passed to tornado's HTTPRequest class e.g. request_timeout
+
+        # extra kwargs passed to tornado's HTTPRequest class e.g. request_timeout
+        self.http_request_kwargs = http_request_kwargs or {}
```
Comment (sxtomerg) [inline tornadoes/__init__.py:23]: good renaming :)

**Review by sxtomerg — CHANGES_REQUESTED** [review]

File: tornadoes/__init__.py (hunk 3 above)
Comment (Matityahul) [reply tornadoes/__init__.py:52]: you understand that's not my code right?
There are a lot more things I would change in this code but I changed almost only what I must changed.
But anyway, I don't think it's so bad the way they wrote it, I would prefer it by your way.

File: tornadoes/__init__.py (hunk 5 above)
Comment (Matityahul) [reply tornadoes/__init__.py:49]: I can't change parameter names, I want to suggest my changes as a pull-request to the original repository too, I can't change the API...

File: tornadoes/__init__.py (hunk 4 above)
Comment (Matityahul) [reply tornadoes/__init__.py:26]: I always write private methods before public methods. I know what you think about the order of method but I don't agree with you.

File: tornadoes/__init__.py (hunk 1 above)
Comment (tomer-senexx) [reply tornadoes/__init__.py:13]: Very nice, but just note that usually class variables should be in lower case

**Review by sxtomerg — APPROVED** [review]

File: tornadoes/__init__.py [hunk 9]
```diff
@@ -10,62 +10,65 @@
...
 
     @staticmethod
-    def from_uri(uri, io_loop=None, custom_client=None):
+    def _create_query_string(params):
+        """
+        Support Elasticsearch 5.X
```
Comment (tomer-senexx) [inline tornadoes/__init__.py:28]: Please add one more comment about what you are trying to do here, why is it necessary to ES 5..

**Pull request merged**